src/uct/cuda/cuda_copy/cuda_copy_md.c

```diff
@@ -318,24 +324,47 @@ uct_cuda_copy_md_query_attributes(uct_cuda_copy_md_t *md, const void *address,
 
     cu_err = cuPointerGetAttributes(ucs_static_array_size(attr_data), attr_type,
                                     attr_data, (CUdeviceptr)address);
-    if ((cu_err != CUDA_SUCCESS) || (cuda_mem_mype != CU_MEMORYTYPE_DEVICE)) {
+    if (cu_err != CUDA_SUCCESS) {
         /* pointer not recognized */
         return UCS_ERR_INVALID_ADDR;
     }
 
-    if (is_managed || (cuda_mem_ctx == NULL)) {
+#if HAVE_CUDA_FABRIC
+    /* Check if memory is allocated using VMM API and see if host memory needs
+     * to be treated as pinned device memory */
+    cu_err = cuMemRetainAllocationHandle(&alloc_handle, (void*)address);
+    if (cu_err == CUDA_SUCCESS) {
+        is_vmm_allocation = 1;
+        if (cuda_mem_type == CU_MEMORYTYPE_HOST) {
+            cu_err = cuMemGetAllocationPropertiesFromHandle(&prop, alloc_handle);
+            if (cu_err != CUDA_SUCCESS) {
+                cuMemRelease(alloc_handle);
+                return UCS_ERR_INVALID_ADDR;
+            }
+
+            /* if location type is NUMA or NUMA_CURRENT the memory is EGM and can be
+             * potentially accessed by copy engines */
+            if ((prop.location.type == CU_MEM_LOCATION_TYPE_HOST_NUMA) ||
+                (prop.location.type == CU_MEM_LOCATION_TYPE_HOST_NUMA_CURRENT)) {
+                cuda_mem_type = CU_MEMORYTYPE_DEVICE;
+            } else {
+                cuMemRelease(alloc_handle);
+                return UCS_ERR_INVALID_ADDR;
+            }
+        }
+        cuMemRelease(alloc_handle);
+    }
```

Comment:
Please move the logic that checks if host memory needs to be treated as device memory to a separate function.